### PR TITLE
Refactors `LLMS_Product::is_purchaseable()` to fix logical errors

### LINKED ISSUE #1121: Enrollment End date rule is not obeyed
### Reproduction Steps

On a course, Go to Restrictions
Put any start date
Put an end date(like yesterday or two days ago)

### Expected Behavior

Should not allow users enroll for the course.

### Actual Behavior

Users can still enroll even after enrollment date has lapsed.

### Error Messages / Logs

HS-119283

### PR DESCRIPTION
**Pull request by thomasplevy**
## Description

This adds new methods to allow fix logical errors in the pricing table template which causes the issue identified in linked issue.

Fixes #1121 

This also fixes various pieces of documentation throughout the updated files.

## How has this been tested?

New tests were written prior to fixing the error which would expose the issue encountered in #1121, the code was then refactored to fix the failing tests

A lot of new tests were added as well

I've also manually tested viewing the pricing table in various scenarios to ensure no regressions have been created.

## Types of changes

Adds new methods and updates the template to fix the logical flaws.

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. 
- [x] My code follows the LifterLMS Coding & Documentation Standards.

### CODE REVIEW

File: includes/models/model.llms.product.php [hunk 1]
```diff
@@ -210,6 +242,52 @@ public function get_pricing_table_columns_count( $free_only = false ) {
 		return apply_filters( 'llms_get_product_pricing_table_columns_count', $cols, $this, $count, $free_only );
 	}
 
+	/**
+	 * Retrieve a list of restrictions on the product
+	 *
+	 * Restrictions are used to in conjunction with "is_purchaseable()" to
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.product.php:248]:
```suggestion
	 * Restrictions are used to in conjunction with "is_purchasable()" to
```

File: includes/models/model.llms.product.php [hunk 2]
```diff
@@ -244,26 +351,20 @@ public function has_free_access_plan() {
 	 */
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.product.php:351]: missing changelog here, on purpose?

File: templates/product/pricing-table.php [hunk 3]
```diff
@@ -2,66 +2,103 @@
...
-			<?php elseif ( $course->has_date_passed( 'enrollment_end_date' ) ) : ?>
-				<?php llms_print_notice( $course->get( 'enrollment_closed_message' ), 'error' ); ?>
-			<?php endif; ?>
-		<?php endif; ?>
-		<?php if ( ! $course->has_capacity() ) : ?>
-			<?php llms_print_notice( $course->get( 'capacity_message' ), 'error' ); ?>
-		<?php endif; ?>
-	<?php endif; ?>
+		/**
+		 * Pricing table output when the user is not enrolled but the product is not purchaseable.
```
Comment (eri-trabiccolo) [inline templates/product/pricing-table.php:93]:
```suggestion
		 * Pricing table output when the user is not enrolled but the product is not purchasable.
```

File: templates/product/pricing-table.php [hunk 4]
```diff
@@ -2,66 +2,103 @@
...
+		 *
+		 * Hooked: llms_template_product_not_purchaseable - 10
```
Comment (eri-trabiccolo) [inline templates/product/pricing-table.php:95]:
```suggestion
		 * Hooked: llms_template_product_not_purchasable - 10
```

File: tests/phpunit/unit-tests/functions/class-llms-test-functions-templates-pricing-table.php [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+	}
+
+	/**
+	 * Test lifterlms_template_pricing_table(): paid plan with no enabled gateways.
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	public function test_lifterlms_template_pricing_table_not_purchaseable() {
```
Comment (eri-trabiccolo) [inline tests/phpunit/unit-tests/functions/class-llms-test-functions-templates-pricing-table.php:91]:
```suggestion
	public function test_lifterlms_template_pricing_table_not_purchasable() {
```

File: includes/models/model.llms.product.php (hunk 2 above)
Comment (thomasplevy) [reply includes/models/model.llms.product.php:351]: obviously I didn't miss it on purpose. :-p

**Pull request merged**